### PR TITLE
Fix `<em>` issue with mixed content #1410

### LINKED ISSUE #1410: Incompatibility between links wrapped in asterisks, pedantic, and GFM
**Describe the bug**

I have a named link, that I want wrapped in asterisks, essentially making it italic.

**To Reproduce**

```markdown
- italic
  - [*named link*][some-url]
  - *[named link][some-url]*
  - [_named link_][some-url]
  - _[named link][some-url]_
- bold
  - [**named link**][some-url]
  - **[named link][some-url]**
  - [__named link__][some-url]
  - __[named link][some-url]__
- bold italic
  - [***named link***][some-url]
  - ***[named link][some-url]***
  - [___named link___][some-url]
  - ___[named link][some-url]___
  - [*__named link__*][some-url]
  - [__*named link*__][some-url]
  - __*[named link][some-url]*__
- code
  - [`named link`][some-url]
- code italic
  - *[`named link`][some-url]*
  - [*`named link`*][some-url]
  - _[`named link`][some-url]_
  - [_`named link`_][some-url]
- code bold
  - **[`named link`][some-url]**
  - [**`named link`**][some-url]
  - __[`named link`][some-url]__
  - [__`named link`__][some-url]
- code bold italic
  - [***`named link`***][some-url]
  - ***[`named link`][some-url]***
  - [___`named link`___][some-url]
  - ___[`named link`][some-url]___
  - [*__`named link`__*][some-url]
  - [__*`named link`*__][some-url]
  - __*[`named link`][some-url]*__
```

The offending lines are:

```markdown
- *[named link][some-url]*
- ***[named link][some-url]***
- __*[named link][some-url]*__
- *[`named link`][some-url]*
- ***[`named link`][some-url]***
- __*[`named link`][some-url]*__
```

Enabling the `pedantic` option fixes this issue, but at the same time it breaks lots of other stuff, like GFM tables and strikethrough. Also note that I only tested with named links, but it may be broken for other types of links as well.

**Expected behavior**

Well, I'm looking at the exact same example using Remark - another popular parser, that I think was forked out of Marked back then, and everything renders correctly.

Let me know if I need to provider more information.

### PR DESCRIPTION
**Pull request by x13machine**
**Marked version:**

**Markdown flavor:** Markdown.pl|CommonMark|GitHub Flavored Markdown|n/a

## Description

- Fixes this bug #1410 
- I added a test case for this bug.

It was caused `_` and `*` parsed being differently in the em regex. I just replaced the part that searches for `*test*` and replaced it with the part searches for `_test_` and then just charged the characters. I don't know why they were different. Maybe there's a reason for it. All the tests pass, so I dunno.

## Contributor

- [x] Test(s) exist to ensure functionality and minimize regression

## Committer

In most cases, this should be a different person than the contributor.

- [x] Draft GitHub release notes have been updated.
- [x] CI is green (no forced merge required).
- [x] Merge PR

### CODE REVIEW

**Comment by x13machine** [comment]
ran wrong test command

**Comment by x13machine** [comment]
Wait, is it expecting a bug to be there?

**Comment by UziTech** [comment]
Ya the spec tests that are currently failing will fail if they start passing. You will need to remove `shouldFail: true` from [/test/specs/commonmark/commonmark.0.28.json line #3730](https://github.com/markedjs/marked/blob/master/test/specs/commonmark/commonmark.0.28.json#L3730)

It looks like you are also failing some tests that used to pass.

**Comment by UziTech** [comment]
The `shouldFail` lets us know which tests are fixed by each PR and let us keep track of the spec tests that we need to work on.

**Comment by x13machine** [comment]
ok I figured out why they are different. It was to prevent it from being interpreted as an ordered List. 

Also the underscore group has a bug: https://marked.js.org/demo/?text=_%3Cimg%20src%3D%22https%3A%2F%2Fwww.google.com%2Ffavicon.ico%22%20title%3D%22*_%22%2F%3E%0A%0A*%3Cimg%20src%3D%22https%3A%2F%2Fwww.google.com%2Ffavicon.ico%22%20title%3D%22*%22%2F%3E&options=&version=master

The asterisk was processed correctly. While the underscore was not. So the bug was transferred over.

So the underscore and asterisk groups need to be rewritten.

**Comment by x13machine** [comment]
Ok I think, I fixed it this time.

File: lib/marked.js [hunk 1]
```diff
@@ -8,8 +8,8 @@
 'use strict';
 
 /**
- * Block-Level Grammar
- */
+   * Block-Level Grammar
+   */
```
Comment (UziTech) [inline lib/marked.js]: can you remove the whitespace changes?

**Comment by UziTech** [comment]
@davisjam could you check if these regexps are secure from redos?

Here are the new and old derived regexps for `em`:

### New:
```js
/^_([^\s_])_(?!_)|^\*([^\s*<\[])\*(?!\*)|^_([^\s<][\s\S]*?[^\s_])_(?!_|[^\s!"#$%&'()*+,\-.\/:;<=>?@\[^_{|}~])|^_([^\s_<][\s\S]*?[^\s])_(?!_|[^\s!"#$%&'()*+,\-.\/:;<=>?@\[^_{|}~])|^\*([^\s<"][\s\S]*?[^\s\*])\*(?!\*|[^\s!"#$%&'()*+,\-.\/:;<=>?@\[^_{|}~])|^\*([^\s*"<\[][\s\S]*?[^\s])\*(?!\*)/
```

### Old:
```js
/^_([^\s_])_(?!_)|^\*([^\s*"<\[])\*(?!\*)|^_([^\s][\s\S]*?[^\s_])_(?!_|[^\s!"#$%&'()*+,\-.\/:;<=>?@\[^_{|}~])|^_([^\s_][\s\S]*?[^\s])_(?!_|[^\s!"#$%&'()*+,\-.\/:;<=>?@\[^_{|}~])|^\*([^\s"<\[][\s\S]*?[^\s*])\*(?!\*)|^\*([^\s*"<\[][\s\S]*?[^\s])\*(?!\*)/
```

**Comment by UziTech** [comment]
@x13machine could you rebase this so we could get it merged?

**Comment by x13machine** [comment]
Ok I'll do that today

**Comment by x13machine** [comment]
:confused: How do I fix the errors?

**Comment by UziTech** [comment]
looks like example 455 should be `"shouldFail": true` and the `"shouldFail"` line should be removed on example 418 and 477

Also move the new test files to `/test/specs/new/`

File: test/new/em_list_links.html [hunk 2]
```diff
@@ -0,0 +1,55 @@
+<ul>
```
Comment (UziTech) [inline test/new/em_list_links.html]: Could you move this file to `test/specs/new/em_list_links.html`

File: test/new/em_list_links.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
+- italic
```
Comment (UziTech) [inline test/new/em_list_links.md]: Could you move this file to `test/specs/new/em_list_links.md`

**Comment by UziTech** [comment]
I'm not sure why the latest commit didn't trigger travis to run the tests, but I ran it locally and everything passed.

**Comment by UziTech** [comment]
This still doesn't fix `*[link*](url*` but that can be a separate PR.

[marked demo](https://marked.js.org/demo/?text=*%5Blink*%5D%28url*&version=pr%3A1451)
[commonmark demo](https://spec.commonmark.org/dingus/?text=*%5Blink*%5D%28url*)

**Comment by styfle** [comment]
> I'm not sure why the latest commit didn't trigger travis to run the tests, but I ran it locally and everything passed.

GitHub was having issues earlier today https://twitter.com/githubstatus/status/1131242516479709185

**Comment by 203x** [comment]
Fix this and have a new problem

now
`*test*t
*test*`

parse
`<p><em>test*t
*test</em></p>`

right
`<p><em>test</em>t
<em>test</em></p>`

**Comment by UziTech** [comment]
@203x looks like that issue is tracked in https://github.com/markedjs/marked/issues/520

**Pull request merged**